docs/proposals/auth.md

```diff
@@ -0,0 +1,128 @@
+a `TokenReview`.
+
+The configuration will look as such:
+
+- `--authn-ca=/path/to/ca.crt`
+- `--auth-apiserver=https://$KUBE`
+
+### Authorization ###
+
+Authorization will work by matching Heapster concepts to Kubernetes RBAC
+concepts, and then performing `SubjectAccessReview` requests against the
+API server configured in `--auth-apiserver`.
+
+Concepts will map as such:
+
+#### Historical Queries #####
+
+**Verb**: get / **Subresource**: historical-metrics
+
+Namespace and Pod requests map to their respective resources.  Requests to
+get container metrics would map a request to get metrics on the
+corresponding pod.  Requests for node and system container metrics would
+simply map to cluster permissions to get historical-metrics on nodes.
+
+Getting by Pod UID would require permissions to get historical metrics on
+pods in any namespace, since we cannot tell namespace from the pod UID.
+
+Geting pods by a list would require making multiple requests, although for
+the sake of performance, it may be desirable to just merge them into
+`namespace=$NS, resource=pod, name=""` request.
+
+#### Model Queries ####
+
+**Verb**: get / **Subresource**: metrics
+
+These function more or less identically to historical queries, except that
+they don't have to worry about Pod UIDs.
+
+#### Push Creation ####
+
+**Verb**: create / **Subresource**: metrics, unprefixed-metrics
+
+The metrics subresource indicates that a user is allowed to push metrics
+not prefixed with their username.  The unprefixed-metrics subresource
+indicates that a user is allowed to push metrics with no prefix.  Core
+system metrics can never be overriden, however.
+
+In order to determine the resources to use for push metrics, the push
+handler will first determine the full set of resources involved, and will
+then aggregate according to the following rules:
+
+- a request containing only metrics for a single object (e.g. pods,
+  services, etc) will turn into a query against that resource (containers
+  in a pod are considered equivalent to the pod here).
+
+- a request involving multiple objects of a given type (e.g. multiple
+  pods) in a namespace will yield a review for `namespace=$NS,
+  resource=$RES, name=""` for that resource.
+
+- metrics describing a namespace will yield a review against that
+  namespace.
+
+- metrics describing nodes or system containers on a node will yield
+  a review against each node in question.
+
+Additionally it may be valuable to simply detect when a large number of
+namespaces are in play, and combine that all into a single cluster-wide
+review against `namespace="*"` (for cases when a "system" component is
+pushing metrics against all namespaces).
+
+Implementation
+--------------
+
+The authenticator would be inserted as a filter in the go-restful handlers,
+which would then inject authentication information into the request (via
+go-restful request attributes).  Then, each of the different handlers would
```

Comment:
That would be great, except that it doesn't work for push metrics.  It could work for model and historical queries, though.